rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,222 @@
+
+With the increase in the number of HUDI tables, due to a lack of management capabilities, maintenance costs will become higher. This proposal is to implement an independent compaction/clustering Service to manage the Hudi compaction/clustering job.
+
+## Background
+
+In the current implementation, if the HUDI table needs do compact/cluster, it only has three ways:
+
+1. Use inline compaction/clustering, in this mode the job will be block writing job.
+
+2. Using Async compaction/clustering, in this mode the job execute async but also sharing the resource with HUDI to write a job that may affect the stability of job writing, which is not what the user wants to see.
+
+3. Using independent compaction/clustering job is a better way to schedule the job, in this mode the job execute async and do not sharing resources with writing job, but also has some questions:
+   1. The user needs to manually start an async compaction/clustering application, which means that the user needs to maintain two jobs.
+   2. With the increase in the number of HUDI jobs, there is no unified service to manage compaction/clustering jobs (monitor, retry, history, etc...), which will make maintenance costs increase.
+
+With this effort, we want to provide an independent compaction/clustering Service, it will have these abilities:
+
+- Provides a pluggable execution interface that can adapt to multiple execution engines, such as Spark and Flink.
+
+- With the ability to failover, need to be persisted compaction/clustering message.
+
+- Perfect metrics and reuse HoodieMetric expose to the outside.
+
+- Provide automatic failure retry for compaction/clustering job.
+
+## Implementation
+
+![](service.png)
+
+### Client
+
+Register to Service with CompactionConfig/ClusteringConfig when the job starts. When generating a compaction/clustering plan, report to the compaction/clustering Service address specified in HoodieConfig and request the Service to stop the corresponding compaction/clustering job when the Service is not required or rollback compaction/clustering instant.
+
+### Request Handler
+
+Receive the client's request and save the plan in the meta table.
+
+### Meta Table
+
+Meta table has been implemented by using an internal HUDI MOR Table to store the required metadata for the compaction/clustering message. This table will be internal to a dataset and will not be exposed directly to the user to write/modify.
+
+The Meta Table has the following features:
+
+1. Is a HUDI MOR Table
+
+2. Defines its own custom payload HoodieServicePayload
+
+3. Use basePath + instant as recordKey to guarantee unique
+
+4. Follows all the conventions of a HUDI Table
+
+### Scheduler
+
+Periodically scan the Meta Table and submit compaction/clustering job according to user-specified rules, and need to plug-in the execution engine.
+
+### Schema
+
+Being an actual Hoodie Table, the Meta Table needs to have a schema to store the compaction/clustering message. 
```

Comment:
The service itself will be a stand alone Java program which will be running all the time. So it does not have to be a HUDI table and hence there is no metadata table. 

I feel since this is long running Java program, it should be "fast" and hence use an embedded database like rocksdb.

Reply:
The hoodie table is only an implementation for this, I think we can add other database for the service like MySQL / rocksdb.